File: src/dynamic-programming/746/solution.py
class Solution:
    def minCostClimbingStairs(self, cost: list[int]) -> int:
        n = len(cost)
        dp = [0] * (n)
        dp[0] = cost[0]
        dp[1] = cost[1]
        for i in range(2, n):
            dp[i] = min(dp[i-2], dp[i-1]) + cost[i]
        return min(dp[i-1], dp[i])
            
    def minCostClimbingStairs(self, cost: list[int]) -> int:
        n = len(cost)
        a = cost[0]
        b = cost[1]
        for i in range(2, n):
            a, b = b, min(a, b) + cost[i]
        return min([a, b])
            
        
        
s = Solution()
assert s.minCostClimbingStairs([10,15,20]) == 15
assert s.minCostClimbingStairs([1,100,1,1,1,100,1,1,100,1]) == 6

File: src/dynamic-programming/746/solution.cpp
#include <vector>

// cost = [1,100,1,1,1,100,1,1,100,1]  6
// cost = [10,15,20]  15


class Solution {
public:
    int minCostClimbingStairs(std::vector<int>& cost) {
		int n = cost.size();
        std::vector<int> dp(n + 1, 0);
		dp[0] = 0;
		dp[1] = 0;
		for (int i = 2; i <= n; ++i) {
			dp[i] = std::min(dp[i-1] + cost[i-1], dp[i-2] + cost[i-2]);
		}
		return dp[n];
    }
};